aws_lambda_powertools/utilities/idempotency/persistence/redis.py

```diff
@@ -0,0 +1,149 @@
+
+    def _put_record(self, data_record: DataRecord) -> None:
+        item: Dict[str, Any] = {}
+
+        # Redis works with hset to support hashing keys with multiple attributes
+        # See: https://redis.io/commands/hset/
+        item = {
+            "name": data_record.idempotency_key,
+            "mapping": {
+                self.status_attr: data_record.status,
+            },
+        }
+
+        if data_record.in_progress_expiry_timestamp is not None:
+            item["mapping"][self.in_progress_expiry_attr] = data_record.in_progress_expiry_timestamp
+
+        if self.payload_validation_enabled:
+            item["mapping"][self.validation_key_attr] = data_record.payload_hash
+
+        now = datetime.datetime.now()
+        try:
+            # |     LOCKED     |         RETRY if status = "INPROGRESS"                |     RETRY
+            # |----------------|-------------------------------------------------------|-------------> .... (time)
+            # |             Lambda                                              Idempotency Record
+            # |             Timeout                                                 Timeout
+            # |       (in_progress_expiry)                                          (expiry)
+
+            # Conditions to successfully save a record:
+
+            # The idempotency key does not exist:
+            #    - first time that this invocation key is used
+            #    - previous invocation with the same key was deleted due to TTL
+            idempotency_record = self._connection.hgetall(data_record.idempotency_key)
+            if len(idempotency_record) > 0:
+                # record already exists.
+
+                # status is completed, so raise exception because it exists and still valid
+                if idempotency_record[self.status_attr] == STATUS_CONSTANTS["COMPLETED"]:
+                    raise
+
+                # checking if in_progress_expiry_attr exists
+                # if in_progress_expiry_attr exist, must be lower than now
+                if self.in_progress_expiry_attr in idempotency_record and int(
+                    idempotency_record[self.in_progress_expiry_attr],
+                ) > int(now.timestamp() * 1000):
+                    raise
+
+            logger.debug(f"Putting record on Redis for idempotency key: {data_record.idempotency_key}")
+            self._connection.hset(**item)
+            # hset type must set expiration after adding the record
+            # Need to review this to get ttl in seconds
+            # Q: should we replace self.expires_after_seconds with _get_expiry_timestamp? more consistent
+            self._connection.expire(name=data_record.idempotency_key, time=self.expires_after_seconds)
+        except Exception:
+            logger.debug(f"Failed to put record for already existing idempotency key: {data_record.idempotency_key}")
+            raise IdempotencyItemAlreadyExistsError
```

Comment:
I'm creating a local project to validate this logic. Were you able to validate this logic by running a local Lambda (with SAM for example) and see how Idempotency works? I feel this logic is missing something, but I need to inspect more.

Reply:
I tried it locally for basic functionality(a random generator that return same number after idem record hits), it works as I imaged, do you have a specific case I should focus on?